### PR TITLE
mul Zq between types

### PR DESCRIPTION
**Pull request by 3pmile**
**Description**

This PR implements...
- [x] the Mul trait for Zq * Z
- [x] the Mul trait for Zq and Standard types u64, u32, u16, u8, i64, i32, i16, i8

**Testing**

- [x] I added basic working examples (possibly in doc-comment)
- [x] I added tests for large (pointer representation) values
- [x] I triggered all possible errors in my test in every possible way
- [x] I included tests for all reasonable edge cases

**Checklist:**

- [x] I have performed a self-review of my own code
  - [x] The code provides good readability and maintainability s.t. it fulfills best practices like talking code, modularity, ...
    - [x] The chosen implementation is not more complex than it has to be
  - [x] My code should work as intended and no side effects occur (e.g. memory leaks)
  - [x] The doc comments fit our style guide
  - [x] I have credited related sources if needed

### CODE REVIEW

File: src/integer_mod_q/z_q/arithmetic/mul.rs [hunk 1]
```diff
@@ -95,6 +99,49 @@ impl Zq {
...
+        let mut out = fmpz(0);
+        unsafe {
+            fmpz_mod_mul_fmpz(
+                &mut out,
+                &self.value.value,
+                &other.value,
+                &*self.modulus.modulus,
+            );
+        }
+        Zq::from_z_modulus(&Z::from_fmpz(&out), &self.modulus)
```
Comment (MoogSven) [inline src/integer_mod_q/z_q/arithmetic/mul.rs:137]: There two issues with this line:
1. `Zq::from_z_modulus` applies the modulus again (which is not necessary since `fmpz_mod_mul_fmpz` already ensures the reduced form.
2. `out` is never dropped. `Z::from_fmpz` does not take ownership, therefore this function would still be responsible for it.

```suggestion
        Zq {
            modulus: self.modulus.clone(),
            value: Z { value: out },
        }
```

File: src/integer_mod_q/z_q/arithmetic/mul.rs [hunk 2]
```diff
@@ -165,3 +212,157 @@ mod test_mul {
...
+    }
+
+    /// testing multiplication for big numbers
+    #[test]
+    fn mul_large_numbers() {
+        let a: Zq = Zq::try_from((i64::MAX, u64::MAX - 58)).unwrap();
+        let b: Zq = Zq::try_from((i64::MAX - 1, i64::MAX)).unwrap();
+        let c: Z = Z::from(u64::MAX);
+        let d: Zq = a * &c;
+        let e: Zq = b * c;
```
Comment (MoogSven) [inline src/integer_mod_q/z_q/arithmetic/mul.rs:265]:
```suggestion

        let d: Zq = a * &c;
        let e: Zq = b * c;
        
```

File: src/integer_mod_q/z_q/arithmetic/mul.rs [hunk 3]
```diff
@@ -165,3 +212,157 @@ mod test_mul {
...
+        let a: Zq = Zq::try_from((4, 11)).unwrap();
+        let b: u64 = 1;
+        let c: u32 = 1;
+        let d: u16 = 1;
+        let e: u8 = 1;
+        let f: i64 = 1;
+        let g: i32 = 1;
+        let h: i16 = 1;
+        let i: i8 = 1;
+        let _: Zq = &a * &b;
```
Comment (MoogSven) [inline src/integer_mod_q/z_q/arithmetic/mul.rs:296]:
```suggestion
        let i: i8 = 1;
        
        let _: Zq = &a * &b;
```
